Wrong spring-modulith-observability dependency for Spring Modulith 2.1.0

the configured spring modulith + open telemetry + actuator results in a build that doesnt compile 

here's the [configuration](https://start.spring.io/#!type=maven-project&language=java&platformVersion=4.1.0&packaging=jar&configurationFileFormat=properties&jvmVersion=25&groupId=com.example&artifactId=demo&packageName=com.example.demo&dependencies=actuator,opentelemetry,modulith)

I get the following error:

```
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/modulith/spring-modulith-observability/2.1.0/spring-modulith-observability-2.1.0.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.859 s
[INFO] Finished at: 2026-06-16T12:50:30+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project demo: Could not resolve dependencies for project com.example:demo:jar:0.0.1-SNAPSHOT
[ERROR] dependency: org.springframework.modulith:spring-modulith-observability:jar:2.1.0 (runtime)
[ERROR] 	Could not find artifact org.springframework.modulith:spring-modulith-observability:jar:2.1.0 in central (https://repo.maven.apache.org/maven2)
[ERROR] 
[ERROR] -> [Help 1]
```

--- a/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizer.java
+++ b/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizer.java
@@ -25,6 +25,9 @@ import io.spring.initializr.generator.buildsystem.Dependency.Builder;
 import io.spring.initializr.generator.buildsystem.DependencyContainer;
 import io.spring.initializr.generator.buildsystem.DependencyScope;
 import io.spring.initializr.generator.spring.build.BuildCustomizer;
+import io.spring.initializr.generator.version.Version;
+import io.spring.initializr.generator.version.VersionParser;
+import io.spring.initializr.generator.version.VersionRange;
 
 /**
  * Registers Spring Modulith dependencies to the build file of the project to be created,
@@ -37,25 +40,27 @@ import io.spring.initializr.generator.spring.build.BuildCustomizer;
  */
 class SpringModulithBuildCustomizer implements BuildCustomizer<Build> {
 
-	private static final Collection<String> OBSERVABILITY_DEPENDENCIES = List.of("actuator", "datadog", "graphite",
-			"influx", "new-relic", "otlp-metrics", "prometheus", "wavefront", "zipkin");
+	private static final VersionRange SPRING_BOOT_4_1_OR_LATER = VersionParser.DEFAULT.parseRange("4.1.0-M1");
+
+	private static final Collection<String> OBSERVABILITY_DEPENDENCIES = List.of("datadog", "graphite", "influx",
+			"new-relic", "otlp-metrics", "prometheus", "wavefront", "zipkin");
 
 	private static final Collection<String> PERSISTENCE = List.of("jdbc", "jpa", "mongodb", "neo4j");
 
 	private static final Collection<String> BROKERS = List.of("activemq", "amqp", "artemis", "kafka");
 
+	private final Version version;
+
+	SpringModulithBuildCustomizer(Version version) {
+		this.version = version;
+	}
+
 	@Override
 	public void customize(Build build) {
 		DependencyContainer dependencies = build.dependencies();
-		if (dependencies.has("actuator")) {
-			dependencies.add("modulith-actuator", modulithDependency("actuator").scope(DependencyScope.RUNTIME));
-		}
+		addActuatorAndObservabilityDependencies(dependencies);
 		if (dependencies.has("flyway")) {
 			dependencies.add("modulith-runtime", modulithDependency("runtime").scope(DependencyScope.RUNTIME));
-		}
-		if (OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has)) {
-			dependencies.add("modulith-observability",
-					modulithDependency("observability").scope(DependencyScope.RUNTIME));
 		}
 		addEventPublicationRegistryBackend(build);
 		if (addEventExternalizationDependency(build)) {
@@ -63,6 +68,30 @@ class SpringModulithBuildCustomizer implements BuildCustomizer<Build> {
 		}
 		dependencies.add("modulith-starter-test",
 				modulithDependency("starter-test").scope(DependencyScope.TEST_COMPILE));
+	}
+
+	private void addActuatorAndObservabilityDependencies(DependencyContainer dependencies) {
+
+		if (dependencies.has("actuator") && OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has)) {
+			dependencies.add("modulith-starter-insight",
+					modulithDependency("starter-insight").scope(DependencyScope.COMPILE));
+			return;
+		}
+		if (dependencies.has("actuator")) {
+			dependencies.add("modulith-actuator", modulithDependency("actuator").scope(DependencyScope.RUNTIME));
+		}
+		if (dependencies.has("actuator") || OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has)) {
+			if (SPRING_BOOT_4_1_OR_LATER.match(this.version)) {
+				dependencies.add("modulith-observability-api",
+						modulithDependency("observability-api").scope(DependencyScope.COMPILE));
+				dependencies.add("modulith-observability-core",
+						modulithDependency("observability-core").scope(DependencyScope.RUNTIME));
+			}
+			else {
+				dependencies.add("modulith-observability",
+						modulithDependency("observability").scope(DependencyScope.RUNTIME));
+			}
+		}
 	}
 
 	private boolean addEventPublicationRegistryBackend(Build build) {

--- a/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizer.java
+++ b/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizer.java
@@ -71,17 +71,16 @@ class SpringModulithBuildCustomizer implements BuildCustomizer<Build> {
 	}
 
 	private void addActuatorAndObservabilityDependencies(DependencyContainer dependencies) {
-
-		if (dependencies.has("actuator") && OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has)) {
+		if (hasActuator(dependencies) && hasObservability(dependencies)) {
 			dependencies.add("modulith-starter-insight",
 					modulithDependency("starter-insight").scope(DependencyScope.COMPILE));
 			return;
 		}
-		if (dependencies.has("actuator")) {
+		if (hasActuator(dependencies)) {
 			dependencies.add("modulith-actuator", modulithDependency("actuator").scope(DependencyScope.RUNTIME));
 		}
-		if (dependencies.has("actuator") || OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has)) {
-			if (SPRING_BOOT_4_1_OR_LATER.match(this.version)) {
+		if (hasActuator(dependencies) || hasObservability(dependencies)) {
+			if (isBoot41orLater()) {
 				dependencies.add("modulith-observability-api",
 						modulithDependency("observability-api").scope(DependencyScope.COMPILE));
 				dependencies.add("modulith-observability-core",
@@ -92,6 +91,18 @@ class SpringModulithBuildCustomizer implements BuildCustomizer<Build> {
 						modulithDependency("observability").scope(DependencyScope.RUNTIME));
 			}
 		}
+	}
+
+	private boolean isBoot41orLater() {
+		return SPRING_BOOT_4_1_OR_LATER.match(this.version);
+	}
+
+	private boolean hasActuator(DependencyContainer dependencies) {
+		return dependencies.has("actuator");
+	}
+
+	private boolean hasObservability(DependencyContainer dependencies) {
+		return OBSERVABILITY_DEPENDENCIES.stream().anyMatch(dependencies::has);
 	}
 
 	private boolean addEventPublicationRegistryBackend(Build build) {

--- a/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithProjectGenerationConfiguration.java
+++ b/start-site/src/main/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithProjectGenerationConfiguration.java
@@ -17,6 +17,7 @@
 package io.spring.start.site.extension.dependency.springmodulith;
 
 import io.spring.initializr.generator.condition.ConditionalOnRequestedDependency;
+import io.spring.initializr.generator.project.ProjectDescription;
 import io.spring.initializr.generator.project.ProjectGenerationConfiguration;
 
 import org.springframework.context.annotation.Bean;
@@ -32,8 +33,8 @@ import org.springframework.context.annotation.Bean;
 class SpringModulithProjectGenerationConfiguration {
 
 	@Bean
-	SpringModulithBuildCustomizer springModulithBuildCustomizer() {
-		return new SpringModulithBuildCustomizer();
+	SpringModulithBuildCustomizer springModulithBuildCustomizer(ProjectDescription description) {
+		return new SpringModulithBuildCustomizer(description.getPlatformVersion());
 	}
 
 }

--- a/start-site/src/test/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizerTests.java
+++ b/start-site/src/test/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizerTests.java
@@ -23,6 +23,7 @@ import io.spring.initializr.generator.buildsystem.maven.MavenBuild;
 import io.spring.initializr.generator.version.Version;
 import io.spring.initializr.metadata.InitializrMetadata;
 import io.spring.initializr.metadata.support.MetadataBuildItemResolver;
+import io.spring.start.site.SupportedBootVersion;
 import io.spring.start.site.extension.AbstractExtensionTests;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
@@ -35,31 +36,36 @@ import static org.assertj.core.api.Assertions.assertThat;
  *
  * @author Oliver Drotbohm
  * @author Eddú Meléndez
+ * @author Moritz Halbritter
  */
 class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 
-	private final SpringModulithBuildCustomizer customizer = new SpringModulithBuildCustomizer(Version.parse("4.1.0"));
-
 	@Test
 	void registersTestAndCoreStarterWhenModulithIsSelected() {
-		Build build = createBuild("modulith");
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith");
 		assertThat(build.dependencies().ids()).contains("modulith");
 		assertThat(build.dependencies().ids()).contains("modulith-starter-test");
 	}
 
 	@Test
 	void registersActuatorStarterIfActuatorsIsPresent() {
-		Build build = createBuild("modulith", "actuator");
-		this.customizer.customize(build);
-		assertThat(build.dependencies().ids()).contains("modulith-actuator");
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", "actuator");
+		assertThat(build.dependencies().ids()).contains("modulith-actuator", "modulith-observability-api",
+				"modulith-observability-core");
 		assertThat(build.dependencies().ids()).doesNotContain("modulith-starter-insight");
 	}
 
 	@Test
+	void registersActuatorStarterIfActuatorsIsPresentForOlderBootVersion() {
+		Build build = createBuild(SupportedBootVersion.V4_0, "modulith", "actuator");
+		assertThat(build.dependencies().ids()).contains("modulith-actuator", "modulith-observability");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-observability-api",
+				"modulith-observability-core", "modulith-starter-insight");
+	}
+
+	@Test
 	void registersRuntimeIfFlywayIsPresent() {
-		Build build = createBuild("modulith", "flyway");
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", "flyway");
 		assertThat(build.dependencies().ids()).contains("modulith-runtime");
 	}
 
@@ -67,9 +73,7 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 	@ValueSource(strings = { "actuator", "datadog", "graphite", "influx", "new-relic", "otlp-metrics", "prometheus",
 			"wavefront", "zipkin" })
 	void registersObservabilityStarterIfObservabilityDependencyIsPresent(String dependency) {
-		Build build = createBuild("modulith");
-		build.dependencies().add(dependency);
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", dependency);
 		assertThat(build.dependencies().ids()).contains("modulith-observability-api", "modulith-observability-core");
 		assertThat(build.dependencies().ids()).doesNotContain("modulith-starter-insight");
 	}
@@ -77,9 +81,7 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 	@ParameterizedTest
 	@ValueSource(strings = { "jdbc", "jpa", "mongodb", "neo4j" })
 	void presenceOfSpringDataModuleAddsModuleEventStarter(String store) {
-		Build build = createBuild("modulith");
-		build.dependencies().add("data-" + store);
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", "data-" + store);
 		assertThat(build.dependencies().ids()).contains("modulith-starter-" + store);
 		assertThat(build.dependencies().ids()).doesNotContain("modulith-starter-core");
 	}
@@ -87,8 +89,7 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 	@ParameterizedTest
 	@ValueSource(strings = { "amqp", "kafka" })
 	void addsExternalizationDependency(String broker) {
-		Build build = createBuild("modulith", broker);
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", broker);
 		assertThat(build.dependencies().ids()).contains("modulith-events-" + broker);
 		assertThat(build.dependencies().ids()).contains("modulith-events-api");
 	}
@@ -96,41 +97,42 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 	@ParameterizedTest
 	@ValueSource(strings = { "activemq", "artemis" })
 	void addsJmsExternalizationDependency(String broker) {
-		Build build = createBuild("modulith", broker);
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", broker);
 		assertThat(build.dependencies().ids()).contains("modulith-events-jms");
 		assertThat(build.dependencies().ids()).contains("modulith-events-api");
 	}
 
 	@Test
 	void addsInsightStarterIfBothActuatorAndObservabilityDependenciesDeclared() {
-
-		Build build = createBuild("modulith", "actuator", "otlp-metrics");
-		this.customizer.customize(build);
+		Build build = createBuild(SupportedBootVersion.V4_1, "modulith", "actuator", "otlp-metrics");
 		assertThat(build.dependencies().ids()).contains("modulith-starter-insight");
 		assertThat(build.dependencies().ids()).doesNotContain("modulith-actuator", "modulith-observability-api",
 				"modulith-observability-core");
 	}
 
 	@Test
-	void addsLegacyObservabilityDependenciesForOlderBootVersion() {
+	void addsInsightStarterIfBothActuatorAndObservabilityDependenciesDeclaredForOlderBootVersion() {
+		Build build = createBuild(SupportedBootVersion.V4_0, "modulith", "actuator", "otlp-metrics");
+		assertThat(build.dependencies().ids()).contains("modulith-starter-insight");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-actuator", "modulith-observability",
+				"modulith-observability-api", "modulith-observability-core");
+	}
 
-		Build build = createBuild("modulith", "otlp-metrics");
-		new SpringModulithBuildCustomizer(Version.parse("4.0.0")).customize(build);
+	@Test
+	void addsLegacyObservabilityDependenciesForOlderBootVersion() {
+		Build build = createBuild(SupportedBootVersion.V4_0, "modulith", "otlp-metrics");
 		assertThat(build.dependencies().ids()).contains("modulith-observability");
 		assertThat(build.dependencies().ids()).doesNotContain("modulith-observability-api",
 				"modulith-observability-core", "modulith-starter-insight");
 	}
 
-	private Build createBuild(String... dependencies) {
+	private Build createBuild(SupportedBootVersion springBootVersion, String... dependencies) {
 		InitializrMetadata metadata = getMetadata();
-		MavenBuild build = new MavenBuild(new MetadataBuildItemResolver(metadata, getDefaultPlatformVersion(metadata)));
+		Version parsedVersion = Version.parse(springBootVersion.getVersion());
+		MavenBuild build = new MavenBuild(new MetadataBuildItemResolver(metadata, parsedVersion));
 		Arrays.stream(dependencies).forEach(build.dependencies()::add);
+		new SpringModulithBuildCustomizer(parsedVersion).customize(build);
 		return build;
-	}
-
-	private Version getDefaultPlatformVersion(InitializrMetadata metadata) {
-		return Version.parse(metadata.getBootVersions().getDefault().getId());
 	}
 
 }

--- a/start-site/src/test/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizerTests.java
+++ b/start-site/src/test/java/io/spring/start/site/extension/dependency/springmodulith/SpringModulithBuildCustomizerTests.java
@@ -38,7 +38,7 @@ import static org.assertj.core.api.Assertions.assertThat;
  */
 class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 
-	private final SpringModulithBuildCustomizer customizer = new SpringModulithBuildCustomizer();
+	private final SpringModulithBuildCustomizer customizer = new SpringModulithBuildCustomizer(Version.parse("4.1.0"));
 
 	@Test
 	void registersTestAndCoreStarterWhenModulithIsSelected() {
@@ -53,6 +53,7 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 		Build build = createBuild("modulith", "actuator");
 		this.customizer.customize(build);
 		assertThat(build.dependencies().ids()).contains("modulith-actuator");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-starter-insight");
 	}
 
 	@Test
@@ -69,7 +70,8 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 		Build build = createBuild("modulith");
 		build.dependencies().add(dependency);
 		this.customizer.customize(build);
-		assertThat(build.dependencies().ids()).contains("modulith-observability");
+		assertThat(build.dependencies().ids()).contains("modulith-observability-api", "modulith-observability-core");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-starter-insight");
 	}
 
 	@ParameterizedTest
@@ -98,6 +100,26 @@ class SpringModulithBuildCustomizerTests extends AbstractExtensionTests {
 		this.customizer.customize(build);
 		assertThat(build.dependencies().ids()).contains("modulith-events-jms");
 		assertThat(build.dependencies().ids()).contains("modulith-events-api");
+	}
+
+	@Test
+	void addsInsightStarterIfBothActuatorAndObservabilityDependenciesDeclared() {
+
+		Build build = createBuild("modulith", "actuator", "otlp-metrics");
+		this.customizer.customize(build);
+		assertThat(build.dependencies().ids()).contains("modulith-starter-insight");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-actuator", "modulith-observability-api",
+				"modulith-observability-core");
+	}
+
+	@Test
+	void addsLegacyObservabilityDependenciesForOlderBootVersion() {
+
+		Build build = createBuild("modulith", "otlp-metrics");
+		new SpringModulithBuildCustomizer(Version.parse("4.0.0")).customize(build);
+		assertThat(build.dependencies().ids()).contains("modulith-observability");
+		assertThat(build.dependencies().ids()).doesNotContain("modulith-observability-api",
+				"modulith-observability-core", "modulith-starter-insight");
 	}
 
 	private Build createBuild(String... dependencies) {

